docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+
+```
+
+</details>
+
+
+#### Request tracking
+
+This will hopefully be handled by the LSP library. Although we might need to be able to keep track of requests in progress for the purpose of cancellation.
+
+### NewVsix
+
+New VS extension that will be backed by the LSP server. It should mostly contain plumbing to connect LSP to the editor and CPS.
+
+> Is there a library available to simplify some of this?
+
+> Will we be able to build this on top of the new VisualStudio.Extensibility?
+
+It should be able to run side by side with the current extension gradually take over the functionality we will add to LSP.
+
+Depending on the extensibility model we will either use the FsLSPServer executable or just load the library directly.
+
+### FCS
+
+The only strict requirement on FCS is to be able to process open files from memory and non-open files from filesystem. But probably best to abstract that within FsLSP and just provide `ISourceText` objects for both cases.
+
+At the moment this is possible via the experimental `DocumentSource.Custom` parameter. But hopefully we will be able to use a new `FSharpProjectSnapshot` model (from [#15179](https://github.com/dotnet/fsharp/pull/15179)) which combines all the necessary inputs into a single object.
+
+
+## LSP library
+
+The library needs to support all the features we want to implement including the ability to add custom F# ones. It also shouldn't incur an unnecessary performance overhead.
+
+### Candidates
+
+| Library  | Pros | Cons |
+| ---  | --- | --- |
+| We write our own | + Complete control over everything | - A lot of extra work |
+| [Ionide LanguageServerProtocol](https://github.com/ionide/LanguageServerProtocol) | + Battle tested with Ionide and FSAC + Natural F# API | - Might not be able to fully customize it to our need |
+| [Roslyn CLaSP](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/README.md) | + Should be performant + Future support | - Not stable API - Unnatural C# API |
+| [OmniSharp C# LSP](https://github.com/OmniSharp/csharp-language-server-protocol)  |  | - Unnatural C# API |
+
+It probably makes sense to start with an existing library which should be able to get us to a working product and only then consider writing our own in case it's the only way to enable more functionality or performance.
+
+## LSP client (?)
+
+- Is there any code that can be shared by various clients?
+- Most likely not, since each client will be specific to its platform and project system
+
+
+## Diagnostic mode
+
+We will be offering the **Pull Diagnostics** mode where the client will request diagnostics for a particular document. This should be easier to implement and probably more practical anyway.
+
+We could experiment with **Publish Diagnostics** in combination with "keep checking the whole solution in background" option, if we decide to implement it on the server side. But this is a low priority item.
```

Comment:
FSAC has background checking so this would be a take-back.

Reply:
This is again more about who will control the background checking. It could be the client going through all files when it's idle and calling pull diagnostics (like it works now in VS). Or it could be the server and then we might use push diagnostics for it. 

The pull version seems more flexible to start with but open to discussion.

Reply:
I do believe we should make it configurable - either bg worker to push diagnostics, or on-demand pull diagnostics on per-document basis.

Reply:
Yeah, we can do it like that eventually. But we can have a MVP with just pull diagnostics. I'll update it.